bolt/kv.go

```diff
@@ -233,26 +235,49 @@ func (s *KVStore) Restore(ctx context.Context, r io.Reader) error {
 
 		// Swap and reopen under lock.
 		s.mu.Lock()
-		defer s.mu.Unlock()
-
-		if err := s.db.Close(); err != nil {
+		if err := s.swapAndOpenRestored(); err != nil {
+			s.mu.Unlock()
 			return err
 		}
 
-		// Atomically swap temporary file with current DB file.
-		if err := fs.RenameFileWithReplacement(s.tempPath(), s.path); err != nil {
+		// Now that the DB has been re-opened, release the lock so that the
+		// migrations can be run.
```

Comment:
There's definitely some potential for a race condition here & not a higher level lock. My initial thought was that it wouldn't be much worse than the potential race conditions between the multiple different restore endpoints that have to be used to do a (full) restore - there's the restore KV endpoint, restore SQL endpoint, and endpoints for all of the shards that the CLI orchestrates without any server-side locks.

I think what might be fairly easy to do would be to run the migrations on the restored database before swapping it in. I'll look into that.

Reply:
It wasn't too bad to run the migrations on the restored database file(s) prior to swapping out the running database. I think that should take care of any race conditions related to the restore KV/restore SQL endpoint. It might be nice someday to consolidate the restore endpoints similar to how the single metadata backup endpoint works under a single lock to prevent anything funny from happening in between CLI calls to the restore endpoints.